Memory-aligned SampleBuffer to utilize SSE optimizations

uklotzde opened the pull request:
A lean wrapper class that manages allocation/deallocation of a properly aligned CSAMPLE array for SSE optimizations.

I've already replaced some (but not all) C-style arrays in the code base by this managed class in a separate commit. After all C-style arrays have been eliminated the memory management functions could be moved from SampleUtil to SampleBuffer, hidden in an anonymous namespace as an implementation detail behind the scenes.

I need and will use this class in my NewSoundSourceAPI branch.

rryan commented on src/samplebuffer.cpp:
```diff
@@ -0,0 +1,21 @@
+#include "samplebuffer.h"
```
I have a slight preference to stick everything in the header file for easy inlining. Is there any reason it needs its own object?

rryan replied:
I don't really mind sampleutil.h leaking everywhere.

uklotzde replied:
I don't expect any severe performance impacts by not inlining the constructor/destructor. So I preferred to hide the implementation details. The implementation of alloc/free will be moved to SampleBuffer.cpp anyway after all unmanaged arrays have been eliminated. At least then the constructor and destructor will no longer be inlined.

I added the clear() and fill() operations just for convenience. Their invocation is currently not time critical, they are currently used only for initialization.

rryan commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,90 @@
+    void fill(value_type value);
+
+  private:
+    value_type* m_data;
```
Should this have an aligned attribute to help the compiler?

rryan replied:
Er, I guess ```__attribute__((aligned(x)))``` refers to the pointer not to the pointed-to buffer?

uklotzde replied:
My understanding is
_ALIGN_16 is only needed for allocation of arrays on the stack
_RESTRICT/__restrict__ is ignored for function declarations according to the GCC documentation (" This means you only need to specify __restrict__ in a function definition, rather than in a function prototype as well.")

daschuer commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,90 @@
+
+  public:
+    typedef size_t size_type;
+    typedef CSAMPLE value_type;
```
These two typedefs IMHO just hides implementation details. What is the reason for it? 
Since CSAMPE and size_t are already typedefs for its special content, a new typedef seams to be overdone.

uklotzde replied:
Just a habit, STL style. Keep it as close as possible to the container types of Standard C++ like std::vector. I also like if classes and their types are self-contained with centralized type definitions. Doesn't hurt and makes maintenance easier.

uklotzde replied:
Well, it actually does or did hurt a little bit when using such a member typedef as a return type of a function. But this will be fixed with the new (trailing) return value syntax of C++11.

daschuer replied:
Since this is no template class I would prefer to explicit use CSAMPLE, otherwise we will have a lot of implicit casts which may confuse readers (including me). 
Qt uses int for all loops and size types, thats was why I have picked int when optimized the Sample functions. We should use int here too to avoid signed unsigned castings and 64 bit to 32 bit casts.

uklotzde replied:
Ok, I stripped down SampleBuffer as much as I think is possible and sensible. Did you notice how I changed "size_type" from "size_t" to "int"? Only a single line of code was affected :)

Counter question: Why didn't you use a typedef if 'int' is essential for those loops to get optimized? This would be the perfect place in the code to document that decision. Moreover it would be helpful to keep the optimized loop code portable: The required type may vary between OSes (Windows, Linux, OS X), CPU architectures (32- vs 64-bit), compilers (GCC vs Clang), and even between different versions of the same compiler.

Need a simple example of what happens if you advertise native types in an API instead of typedefs? Have a look at __M4AHACK__ and you will see what I'm talking about ;)

rryan replied:
> Counter question: Why didn't you use a typedef if 'int' is essential for those loops to get optimized?

I don't think int is essential for those loops being optimized -- based on the source Daniel provided describing the MSVC loop vectorizer, unsigned int can cause problems if it does not match the natural architecture width but size_t will always be the full integer width of the architecture.

I think size_t is the ideal type to use since it semantically describes what the variable represents. Pretty much the only spots that this matters are index variables in for loops anyway and I already had to change all those in https://github.com/mixxxdj/mixxx/commit/0cdafa3cc0b5679f9d809740928e20844986564a
once #483 merged.

rryan replied:
> Ok, I stripped down SampleBuffer as much as I think is possible and sensible. Did you notice how I changed "size_type" from "size_t" to "int"? Only a single line of code was affected :)

I don't have strong opinions about these typedefs -- this is one nice benefit.

daschuer replied:
I know the arguments pro typedefs and they are valid. 
But they become less important in the age or refactoring tools and intellisense. 
When a user reads size_type, it can be everything int is somehow well defined. 
In Eclipse you will see nice popups showing the function prototype. getting to know what 
size_type will cost an extra click, loosing the current editing context. 
This argument becomes stronger if we have typedefs for a small scopes like here. 

It may rise the question: This buffer returns a value_type, aha. but I need a buffer for CSAMPLE ... 

It is very unlikely that this calls will change its type but CSAMPLE not.

Such small typedef scopes also required type casts. Every typecast is a candidate for an error. I prefer to keep a values under the same type during the whole lifetime. 

On the other hand I support application wide typedefs. And I like you idea to introduce a global type for all CSAMPLE indexes (like we did for gain values)

daschuer commented:
Cool, thank you. 

There is an option to remove the alignment check from the vectorized loops by using 
__builtin_assume_aligned
I do not know if this is worth the work, but we may extend the sampleutil iterface, in a way that we pass the new SampleBuffer, which guarantees the 16 byte alignment.

uklotzde commented:
__builtin_assume_aligned should only be used for special case member functions of sample buffer. You can only guarantee alignment it if the loop starts at the beginning of the allocated buffer (or at an index with the correct alignment)

uklotzde commented:
__builtin_assume_aligned is simply too dangerous when used in general purpose functions that accept arbitrary pointers as arguments!

daschuer commented:
Yes, thats the point. Most functions like copy3withGain() will always be called for the whole buffer. 
But there is no guarantee yet.  When we pass your new class instead of a CSAMPLE* we can probably guarantee the alignment.

daschuer commented:
By the way: Congratulations to #500 :-)

uklotzde commented:
Good idea, Daniel. I would actually prefer free functions with SampleBuffer as a parameter type as you suggested instead of member functions! SampleBuffer should only be responsible for memory management and nothing more. Even clear() and fill() don't need to be member functions. As already mentioned I just added them for convenience, because they are used for initialization of the (whole) buffer content. An edge case ;)

rryan commented on src/engine/enginemaster.cpp:
```diff
@@ -577,23 +574,23 @@ void EngineMaster::process(const int iBufferSize) {
 	    for (int i = 0; i + 1 < iBufferSize; i += 2) {
-                m_pHead[i] = (m_pHead[i] + m_pHead[i + 1]) / 2;
-                m_pHead[i + 1] = (m_pMaster[i] + m_pMaster[i + 1]) / 2;
+	            m_headBuffer[i] = (m_headBuffer[i] + m_headBuffer[i + 1]) / 2;
```
Hm, is this a tab character? Can you check your editor?

rryan commented on src/engine/enginemaster.h:
```diff
@@ -133,13 +134,8 @@ class EngineMaster : public QObject, public AudioSource {
     }
 
     struct ChannelInfo {
-        ChannelInfo(int index)
```
For these helper structs I prefer to keep their definition all in one place so you can easily look at them as a whole.

rryan commented on src/samplebufferutil.h:
```diff
@@ -0,0 +1,17 @@
+#include "sampleutil.h"
+
+// Fills the whole buffer with zeroes
+inline void clear(SampleBuffer* pBuffer) {
```
I'd rather not have these in the global namespace. Why not attach them to the SampleBuffer object as you had them?

daschuer commented:
> I don't think int is essential for those loops being optimized -- based on the source Daniel provided describing the MSVC loop vectorizer, unsigned int can cause problems if it does not match the natural architecture width but size_t will always be the full integer width of the architecture.

> I think size_t is the ideal type to use since it semantically describes what the variable represents. Pretty much the only spots that this matters are index variables in for loops anyway and I already had to change all those in 0cdafa3
once #483 merged. 

Lets stick with int, I have found a blog that only int will "allways" work, unfortunately I cant find it yet. 
but anyway, the loop patterns provided by gcc and which are part of the gcc unit tests are using int 
https://gcc.gnu.org/projects/tree-ssa/vectorization.html
so it cant be wrong doing the same. 
clang is the same opinion: 
http://blog.llvm.org/2014/11/loop-vectorization-diagnostics-and.html
and finally intel 
https://software.intel.com/sites/products/documentation/doclib/iss/2013/compiler/cpp-lin/GUID-E98D4E0A-9730-425D-A898-3BB4AB9B2330.htm

ywwg commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,73 @@
+
+  public:
+    typedef CSAMPLE value_type;
+    typedef int size_type;
```
I'll just add my own two cents again typedefs like this -- I know it makes refactoring easier but this specific use of typedefs only serves to hide the types being used, which is one of the big benefits of this wonderful language called C++ in the first place.  I won't insist on changing it, but I'd prefer language features be used to make things easier for people calling your code, not easier for your fingers when typing the code :)

(if either type was a templated-to-hell mess of <<>>>>>, I would totally understand)

uklotzde replied:
If I write "1.0f" this is equivalent to "CSAMPLE_PEAK". But the latter is more expressive even if you don't see the value anymore. It is actually a benefit that the concrete value is hidden behind the constant's name. It is good practice to give names to values. So why not to types?

We as developers have to provide these abstractions, giving names to concepts: values, types, functions, algorithms, ... An IDE will never be able to decide what "1.0f" somewhere in the code actually means: peak sample value? max. gain value? ...? Keeping complex (software) systems manageable is all about abstractions.

Another argument: The common width of "int" is still 32-bit even on 64-bit architectures. Are you sure that this will be enough for all time considering multi-channel audio at 192 kHz? We are doomed if we ever need to revert this decision later! On the other hand if you have spent a typedef and consequently used it right from the beginning? Well, nothing easier than that :)

I will remove all typedefs from SampleBuffer.

uklotzde replied:
https://software.intel.com/en-us/articles/about-size-t-and-ptrdiff-t

According to Intel size_t (unsigned) and ptrdiff_t (signed) are the natural choice for "...loop counters, array indexing, size storage and address arithmetic."

Another statement is about portability between 32-/64-bit: "...where use of size_t and ptrdiff_t types provides high performance, possibility to operate large data sizes and portability between different platforms."

My conclusion: ptrdiff_t (signed) and size_t (unsigned) should be used for loop counters, array indexing, size storage and address arithmetic to achieve maximum portability and performance. If you don't want to deal with signed/unsigned issues ptrdiff_t would be the natural choice instead of int. Writing "ptrdiff_t" everywhere in the code is a bit awkward and a custom typedef might help here.

daschuer replied:
Nice article, unfortunately the test results are missing. I just did a test for the following loop with -O3
and it turns out that the results a equivalent.  But anyway, if you can effort the work it is a good idea to introduce a global typedef for CSAMPLE counter than we can test under real conditions. 
```
int main() {
	int max = rand() + 2; 
	for (int k = 0; k < max; ++k) {
		printf("%ld",k);
	}	 
	return 0; 
}
``` 
results for int: 
```
	.cfi_startproc
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	pushq	%rbx
	.cfi_def_cfa_offset 24
	.cfi_offset 3, -24
	subq	$8, %rsp
	.cfi_def_cfa_offset 32
	call	rand
	leal	2(%rax), %ebp
	testl	%ebp, %ebp
	jle	.L2
	xorl	%ebx, %ebx
	.p2align 4,,10
	.p2align 3
.L3:
	movl	%ebx, %edx
	xorl	%eax, %eax
	movl	$.LC0, %esi
	movl	$1, %edi
	addl	$1, %ebx
	call	__printf_chk
	cmpl	%ebx, %ebp
	jg	.L3
.L2:
	addq	$8, %rsp
	.cfi_def_cfa_offset 24
	xorl	%eax, %eax
	popq	%rbx
	.cfi_def_cfa_offset 16
	popq	%rbp
	.cfi_def_cfa_offset 8
	ret
	.cfi_endproc
```
results for size_t
```
	.cfi_startproc
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	pushq	%rbx
	.cfi_def_cfa_offset 24
	.cfi_offset 3, -24
	subq	$8, %rsp
	.cfi_def_cfa_offset 32
	call	rand
	leal	2(%rax), %ebp
	movslq	%ebp, %rbp
	testq	%rbp, %rbp
	je	.L2
	xorl	%ebx, %ebx
	.p2align 4,,10
	.p2align 3
.L3:
	movq	%rbx, %rdx
	xorl	%eax, %eax
	movl	$.LC0, %esi
	movl	$1, %edi
	addq	$1, %rbx
	call	__printf_chk
	cmpq	%rbx, %rbp
	ja	.L3
.L2:
	addq	$8, %rsp
	.cfi_def_cfa_offset 24
	xorl	%eax, %eax
	popq	%rbx
	.cfi_def_cfa_offset 16
	popq	%rbp
	.cfi_def_cfa_offset 8
	ret
	.cfi_endproc
```

daschuer replied:
Here a SSE optimized loop, looks equivalent: 

int: 
```
.L3:
	cvtsi2sd	%edx, %xmm0
	addsd	(%rbx,%rcx,8), %xmm0
	addq	$1, %rcx
	cmpl	%ecx, %eax
	cvttsd2si	%xmm0, %edx
	jg	.L3
.L2:
```

int: 
```
.L3:
	cvtsi2sd	%edx, %xmm0
	addsd	(%rbx,%rcx,8), %xmm0
	addq	$1, %rcx
	cmpq	%rcx, %rax
	cvttsd2si	%xmm0, %edx
	ja	.L3
.L2:
```

daschuer replied:
Probably Qt user signed size type because of thoughts like this: http://www.aristeia.com/Papers/C++ReportColumns/sep95.pdf

An other reason is outlined here: 
http://stackoverflow.com/questions/3473675/negative-array-indexes-in-c
a[-2] is a valid expression 
I wonder what happens if you try to pass 
size_t i = 0xFFFFFFFFFFFFFFFF
a[i]

daschuer replied:
```
	int* a = malloc(1); 
	size_t i = 0xFFFFFFFFFFFFFFFF; 

	printf("%p\n", a); 
	printf("%p\n", &a[1]); 
	printf("%p\n", &a[-1]);
	printf("%p\n", &a[i]);

0xfe9010
0xfe9014
0xfe900c
0xfe900c
```

daschuer replied:
This is probably the reason for the extra ```testq``` call in the size_t loop, I just notice.

uklotzde replied:
If we agree on using only signed types then ptrdiff_t is the right choice for indices, sizes and pointer arithmetics.  It should be defined in util/types.h. Now all we need is the perfect name for this alias!

uklotzde replied:
How about this:
#include <cstddef>

// Signed integer type for all kinds of sizes, (array) indices and pointer
// arithmetics. Its size (32-/64-bit) depends on the CPU architecture.
typedef std::ptrdiff_t SINT;

A short name that is easy to remember and hopefully does not collide with existing names. Uppercase as all other typedefs in util/types.h. 'S' stands for 'Signed' to recall this into memory wherever you use this type ;)

daschuer replied:
It will be always a size of a sample buffer or a index of a sample. 
It is useless once we deal with QT containers, so we should probably be more specific. 
CSAMLPE_SIZE -> bad because missleading 
CSAMPLE_INDEX -> quite long and also missleading because it is used for frames as well. 

SINT works for me.

daschuer commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,77 @@
+#include <algorithm> // std::swap
+
+// A sample buffer with properly aligned memory to enable SSE optimizations.
+// The public SINTerface closely follows that of std::vector.
```
mass renaming: SINTerface
"closely follows that of std::vector": does this add valuable info?  Outdated?

daschuer commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,77 @@
+// A sample buffer with properly aligned memory to enable SSE optimizations.
+// The public SINTerface closely follows that of std::vector.
+//
+// No resize operation is provided SINTentionally for maximum efficiency!
```
SINTentionally

daschuer commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,77 @@
+//     ... copy data from sampleBuffer to tempBuffer...
+//     sampleBuffer.swap(sampleBuffer);
+//
+// The local variable tempBuffer can be omitted if no data needs to be
```
This whole paragraph is outdated and probably misleading.

daschuer replied:
.. because we should never call SampleBuffer tempBuffer(newSize) and swap within the engine code, since malloc can be blocking and is not a no op. So if one tries to resize the buffer it must be a very special issue. 
However, swap() might be nice, to re-factor gain = 1 copies.  We have to be careful that no one has a reference to m_data. 
I think that is worth a comment.

daschuer commented on src/engine/enginemaster.cpp:
```diff
@@ -616,8 +613,9 @@ void EngineMaster::addChannel(EngineChannel* pChannel) {
     pChannelInfo->m_pMuteControl->setButtonMode(ControlPushButton::POWERWINDOW);
-    pChannelInfo->m_pBuffer = SampleUtil::alloc(MAX_BUFFER_LEN);
-    SampleUtil::clear(pChannelInfo->m_pBuffer, MAX_BUFFER_LEN);
+    SampleBuffer(MAX_BUFFER_LEN).swap(pChannelInfo->m_sampleBuffer);
```
Hah, thats the real use for swap() :-)

daschuer commented on src/util/types.h:
```diff
@@ -3,8 +3,13 @@
 #include <climits>
 
+// Signed integer type for all kinds of sizes, array indices and pointer
+// arithmetic. Its size (32-/64-bit) depends on the CPU architecture.
```
// Signed integer type for POT array indices, sizes and pointer
// arithmetic. Its size (32-/64-bit) depends on the CPU architecture.
// This should be uses for all CSAMLE operations 
// since it is fast and allows compiler auto vectorizing. 
// For Qt container operations use just int 

"since it is fast and allows compiler auto vectorizing." verification pending

uklotzde committed:
Introduce SINT signed integer type

uklotzde committed:
Memory-aligned SampleBuffer to utilize SSE optimizations

uklotzde commented:
I'm tired of merging so I rebased this branch and removed all modifications of existing code. It now only contains the new typedef SINT and the new class SampleBuffer.

daschuer commented:
IMHO we can merge this to master. @rryan: any objections?

rryan commented on src/samplebuffer.h:
```diff
@@ -0,0 +1,90 @@
+//
+class SampleBuffer {
+    Q_DISABLE_COPY(SampleBuffer)
+    ;
```
could you put the semicolon on the same line as Q_DISABLE_COPY?

rryan commented:
@daschuer works for me -- LGTM

uklotzde committed:
Revert auto-formatting changes

daschuer commented:
Thank you @uklotzde
